File: FloodFill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if(image[sr][sc] == color){
            return image;
        }
        floodFill(image,sr,sc,image[sr][sc], color);
        return image;
    }
    
    public static void floodFill(int[][] image, int i, int j, int oldColor, int newColor){
        if(i<0 || i>=image.length || j<0 || j>=image[0].length || image[i][j] != oldColor){
            return;
        }
        
        image[i][j] = newColor;
        
        floodFill(image, i+1, j, oldColor, newColor);
        floodFill(image, i-1, j, oldColor, newColor);
        floodFill(image, i, j+1, oldColor, newColor);
        floodFill(image, i, j-1, oldColor, newColor);
        
    }
    
}

File: FloodFill.js
/**
 * @param {number[][]} image
 * @param {number} sr
 * @param {number} sc
 * @param {number} color
 * @return {number[][]}
 */
var floodFill = function(image, sr, sc, color) {
    if(image[sr][sc] === color) return image;
    const flood = (image, sr, sc, originalColor, color) => {
        if(sr < 0 || sr >= image.length || sc < 0 || sc >= image[0].length || image[sr][sc] !== originalColor) {
            return;
        }
        image[sr][sc] = color;
        flood(image, sr + 1, sc, originalColor, color);
        flood(image, sr - 1, sc, originalColor, color);
        flood(image, sr, sc + 1, originalColor, color);
        flood(image, sr, sc - 1, originalColor, color);
    }
    const originalColor = image[sr][sc];
    flood(image, sr, sc, originalColor, color);
    return image;
};
